[display] Do not perserve buffers when display changes

mariogrip opened the pull request:
We should *not* preserve buffers when a display/output was disconnected
or connected

If we do not do this, android will still feed us invalid buffers and/or
compositor thread tries to access destroyed buffers.

Needed for secondscreen to work

mariogrip committed:
[display] Do not perserve buffers when display changes

We should *not* preserve buffers when external display was disconnected
or connected

If we do not do this, android will still feed us invalid buffers and/or
compositor thread tries to access destroyed buffers.

mariogrip commented:
Also want to ring up :phone: @RAOF to checkup if this is indeed correct. I saw you made the change (https://github.com/ubports/mir/commit/b6ea61aea861060ca49d6a826a72a3857c0448f4)

RAOF commented on src/platforms/android/server/display.cpp:
```diff
@@ -359,9 +359,9 @@ bool mga::Display::apply_if_configuration_preserves_display_buffers(
+        return false;
     }
-    return false;
+    return true;
```
Won't this result in the compositor incorrectly thinking that the display configuration has been applied?

RAOF commented on src/platforms/android/server/display.cpp:
```diff
@@ -359,9 +359,9 @@ bool mga::Display::apply_if_configuration_preserves_display_buffers(
     {
         configure_locked(conf, lock);
-        return true;
+        return false;
```
This is definitely wrong; it should return `true` if it applied the configuration. Otherwise the compositor will just call `configure` again.

RAOF reviewed (commented):
I'm pretty sure this is wrong.

If you're having problems, it's safe (but will result in compositor pauses) to do nothing but return `false` from `apply_if_configuration_preserves_display_buffers`

peat-psuwit commented on src/platforms/android/server/display.cpp:
```diff
@@ -359,9 +359,9 @@ bool mga::Display::apply_if_configuration_preserves_display_buffers(
     if (!config.external().connected || displays.display_present(mga::DisplayName::external))
```
To detect if the buffers should be preserved, the condition should read:
```c++
        if (config.external().connected || !displays.display_present(mga::DisplayName::external))
```
i.e. preserve the buffer if the external display is connected, or it's not but we've already dealt with it.


But I would rather change the whole function to read:
```c++
        if (!config.external().connected && displays.display_present(mga::DisplayName::external))
            return false;

        configured_locked(conf, lock);
        return true;
```
i.e. detect when _not_ to preserve the buffer instead.

mariogrip replied:
Yeah i think this is indeed correct, we should return if previous display has been disconnected and call configured_locked otherwise.  

So the *root* of the problem is that we need to restart the compositor when a display has been disconnected. 

@RAOF is this correct?

peat-psuwit reviewed (changes requested):
First of all, think twice before changing the test. In this case, it makes sure it won't invalidate the buffer needlessly. If the change fails, that means your change breaks this assumption.

It seems like it's the external display disconnection logic that's incorrect, not the return code. See the inline comment on how it should be fixed. Also, it would help if you also write a test to catch this.

mariogrip commented:
@peat-psuwit and @RAOF so I corrected the logic to make sure to invalidate the display buffer when a display has been disconnected, and also added tests for both when to invalidate and the existing when not to.

mariogrip commented:
So a new problem has been raised, this does not work at all on the bq m10, the only way to get that to work with a second display is invalidate regardless.

mariogrip commented:
Would there be any side effects of invalidate regardless @RAOF. AFAIK this is what was done before mir 0.26.

RAOF commented:
It is absolutely safe to do nothing but `return false;` from the function.

`apply_if_configuration_preserves_display_buffers` was introduced to make orientation changing smoother. Previously, *any* change to the configuration (such as changing the scale, or orientation, neither of which change the layouts of the `DisplayBuffer`s) would require the compositor to be stopped, the change made, and the the compositor started again, which takes a noticeable time.

You should *definitely* return false from this function if the number of displays has changed.

mariogrip committed:
[display] Add test for when to invalidate display buffer

This adds test to make sure we invalidate the display buffer
when a display has been disconnected or connected.

mariogrip commented:
Ok, so this now works, so what i did is to invalidate buffers when the display has been connected or disconnected, this seems to work perfectly. 

Please re-review @peat-psuwit and @RAOF :)

peat-psuwit commented on tests/unit-tests/platforms/android/server/test_display.cpp:
```diff
@@ -1165,3 +1165,71 @@ TEST_F(Display, does_not_invalidate_display_buffers_when_it_promised_not_to)
+    for (auto const& db : active_dbs)
+    {
+        EXPECT_THAT(db->transformation(), AnyOf(Eq(rotate_inverted), Eq(rotate_none)));
+    }
```
This should NOT be done in this case. The purpose of this section is to make sure memory that isn't supposed to be invalidated won't be free'ed. However, in this case, we expect it to invalidate the buffer, so this will definitely trigger Valgrind.

mariogrip replied:
fixed

peat-psuwit commented on tests/unit-tests/platforms/android/server/test_display.cpp:
```diff
@@ -1165,3 +1165,71 @@ TEST_F(Display, does_not_invalidate_display_buffers_when_it_promised_not_to)
+                {
+                    active_dbs.push_back(&db);
+                });
+        });
```
Also remove this (see the other comment).

mariogrip replied:
fixed

peat-psuwit reviewed (approved):
PR approved, from China! [1] If @RAOF approves, this PR can be merged. Note that I haven't tested it on a real device yet because I have no devices with an external screen capability (yet). [2]

[1] I have a transit flight here. So, meanwhile, limited internet!
[2] I've received the devices from Jan. However, I didn't set them up yet. It seems like the M10 HD has a micro HDMI port. So, if required, I think I can test it on that.

mariogrip committed:
Also invalidate buffers when a virtual display has (dis)connected

RAOF commented on src/platforms/android/server/display.cpp:
```diff
@@ -348,20 +348,21 @@ bool mga::Display::apply_if_configuration_preserves_display_buffers(
+    if ((!config.external().connected && displays.display_present(mga::DisplayName::external)) ||
+       (config.external().connected && !displays.display_present(mga::DisplayName::external)) ||
+       (!config.virt().connected && displays.display_present(mga::DisplayName::virt)) ||
+       (config.virt().connected && !displays.display_present(mga::DisplayName::virt)))
```
I think this might be pre-existing, but do we allow turning off the *internal* panel? If so, might need to check for that.

mariogrip replied:
Added it in my new commit.

RAOF commented on src/platforms/android/server/display.cpp:
```diff
@@ -348,20 +348,21 @@ bool mga::Display::apply_if_configuration_preserves_display_buffers(
-    }
-    return false;
+    if ((!config.external().connected && displays.display_present(mga::DisplayName::external)) ||
+       (config.external().connected && !displays.display_present(mga::DisplayName::external)) ||
```
Code golf: This can be simplified to `config.external().connected == displays.display_present(mga::DisplayName::external)`

mariogrip replied:
Made this a bit more elegant, please see my new commit :)

UniversalSuperBox reviewed (changes requested):
Tests are not passing with the newest commit

RAOF reviewed (approved):
Looks like it will do what it needs to.

mariogrip committed:
Make the detection of dis(connected) screens more elegant

This makes the detection of dis(connected) screens more elegant by
comparing structs containing output connection information

UniversalSuperBox reviewed (approved):
This part of the secondscreen changes works correctly, as evidenced by plugging in and unplugging a display a bunch while unity8 is not running, watching for "invalidate display buffers" in unity-system-compositor's log every time. Even confusing actions like plugging in but unplugging before the buffer change is complete work correctly.
